taichi/runtime/gfx/runtime.cpp

```diff
@@ -502,8 +472,61 @@ void GfxRuntime::launch_kernel(KernelHandle handle, RuntimeContext *host_ctx) {
   }
 
   // Record commands
-  ti_kernel->generate_command_list(current_cmdlist_.get(), args_buffer.get(),
-                                   ret_buffer.get(), any_arrays, textures);
+  const auto &task_attribs = ti_kernel->ti_kernel_attribs().tasks_attribs;
+
+  for (int i = 0; i < task_attribs.size(); ++i) {
+    const auto &attribs = task_attribs[i];
+    auto vp = ti_kernel->get_pipeline(i);
+    const int group_x = (attribs.advisory_total_num_threads +
+                         attribs.advisory_num_threads_per_group - 1) /
+                        attribs.advisory_num_threads_per_group;
+    ResourceBinder *binder = vp->resource_binder();
+    for (auto &bind : attribs.buffer_binds) {
+      if (bind.buffer.type == BufferType::ExtArr) {
+        binder->rw_buffer(0, bind.binding, any_arrays.at(bind.buffer.root_id));
+      } else if (args_buffer && bind.buffer.type == BufferType::Args) {
+        binder->buffer(0, bind.binding, *args_buffer);
+      } else if (ret_buffer && bind.buffer.type == BufferType::Rets) {
+        binder->rw_buffer(0, bind.binding, *ret_buffer);
+      } else {
+        DeviceAllocation *alloc = ti_kernel->get_buffer_bind(bind.buffer);
+        if (alloc) {
+          binder->rw_buffer(0, bind.binding, *alloc);
+        }
+      }
+    }
+
+    for (auto &bind : attribs.texture_binds) {
+      DeviceAllocation texture = textures.at(bind.arg_id);
+      current_cmdlist_->image_transition(texture, ImageLayout::undefined,
+                                         ImageLayout::shader_read);
+      binder->image(0, bind.binding, texture, {});
+    }
+
+    if (attribs.task_type == OffloadedTaskType::listgen) {
+      for (auto &bind : attribs.buffer_binds) {
+        if (bind.buffer.type == BufferType::ListGen) {
+          // FIXME: properlly support multiple list
+          current_cmdlist_->buffer_fill(
+              ti_kernel->get_buffer_bind(bind.buffer)->get_ptr(0),
+              kBufferSizeEntireSize,
+              /*data=*/0);
+          current_cmdlist_->buffer_barrier(
+              *ti_kernel->get_buffer_bind(bind.buffer));
+        }
+      }
+    }
+
+    current_cmdlist_->bind_pipeline(vp);
+    current_cmdlist_->bind_resources(binder);
+    current_cmdlist_->dispatch(group_x);
+    current_cmdlist_->memory_barrier();
+  }
+
+  for (auto &[id, shadow] : any_array_shadows) {
+    current_cmdlist_->buffer_copy(
+        shadow.get_ptr(0), any_arrays.at(id).get_ptr(0), ext_array_size.at(id));
+  }
```

Comment:
If a kernel is called multiple times (like, in a compute graph,) would this introduce extra bandwidth? Probably the behavior should be documented.